### PR TITLE
feat(embodied): add Pi0.6 RECAP pipeline

### PR DESCRIPTION
**Pull request by LiuZhihao2022**
### Description

**Pipeline Overview**

```
┌───────────────┐    ┌───────────────┐    ┌───────────────┐    ┌───────────────┐
│    Step 1     │    │    Step 2     │    │    Step 3     │    │    Step 4     │
│    Compute    │───▶│  Value Model  │───▶│    Compute    │───▶│      CFG      │
│    Returns    │    │      SFT      │    │  Advantages   │    │    Training   │
└───────────────┘    └───────────────┘    └───────────────┘    └───────────────┘
```

This PR implements **RECAP (RL with Experience and Corrections via Advantage-conditioned Policies)** from — the offline policy-optimization pipeline proposed in Physical Intelligence's π₀.₆ work. RECAP computes returns from logged datasets, trains a value model, estimates per-timestep advantages, and fine-tunes the OpenPI (π₀.₅) policy with **Classifier-Free Guidance (CFG)** conditioned on the advantage labels.

The pipeline has four sequential stages, each wired as a standalone entrypoint:

1. **Step 1 – Compute Returns** (`examples/process/compute_returns.py`): computes discounted returns `G_t = r_t + γ·G_{t+1}` in reverse over each trajectory and writes sidecar `meta/returns_{tag}.parquet` files without mutating the original LeRobot dataset.
2. **Step 2 – Value Model SFT** (`examples/sft/train_value_sft.py`): trains a value model (SigLIP2-so400m vision encoder + Gemma3-270M language model + learnable Critic Expert) to predict normalized returns from (image, language) observations. Supervision is a Categorical Value Distribution over 201 bins on `[-1, 0]`.
3. **Step 3 – Compute Advantages** (`examples/process/compute_advantages.py`): uses the trained value model to compute N-step lookahead advantages `A_t = normalize(r_{t:t+N}) + γ^N·V(o_{t+N}) − V(o_t)`, then labels the top `X%` (default 30%) as positive and the rest as negative. Multi-GPU distributed inference supported.
4. **Step 4 – CFG Training** (`examples/sft/train_cfg_sft.py`): fine-tunes OpenPI (pi0.5) with classifier-free guidance — positive samples act as conditional inputs, negative samples as unconditional, with `unconditional_prob` sample-level dropout regularization. At inference, `cfgrl_guidance_scale` amplifies the difference between conditional and unconditional predictions to steer the policy toward high-advantage actions.

**New files added**

- Models
  - `rlinf/models/embodiment/openpi_cfg/` — OpenPI CFG action model wrapping pi0.5 with advantage-conditioned training
  - `rlinf/models/embodiment/value_model/` — value model (SigLIP2 + Gemma3 + Critic Expert), processing, data collator, checkpoint utils
- Datasets
  - `rlinf/data/datasets/cfg/` — value / advantage / mixture datasets, return loaders, value transforms (LeRobot-format)
- Workers
  - `rlinf/workers/value_sft/fsdp_value_sft_worker.py` — FSDP worker for value model SFT
  - `rlinf/workers/cfg/fsdp_cfg_worker.py` — FSDP worker for CFG policy training
- Checkpoint converter
  - `rlinf/utils/ckpt_convertor/fsdp_convertor/config/fsdp_value_model_convertor.yaml` — FSDP→HF converter config for the value model
- Entry points and configs
  - `examples/process/` — `compute_returns.py`, `compute_advantages.py`, `recompute_advantages_from_value_reward.py` (threshold relabeling without GPU re-inference), `visualize_advantage_dataset.py`, launch scripts, and YAML configs
  - `examples/sft/` — `train_value_sft.py`, `train_cfg_sft.py`, launch scripts, and `libero_sft_value.yaml` / `libero_cfg_openpi.yaml` / `model/value.yaml` / `model/value_model.yaml`

### How has this been tested?

An end-to-end test was run that exercises all four pipeline stages in order on LIBERO-10 Task 0 — **Step 1 compute_returns → Step 2 value_sft → Step 3 compute_advantages → Step 4 cfg_sft**. Each stage produces and validates the expected artifact before the next stage consumes it.

**Step 2 — Value Model SFT**

Value model loss curve:

value model训练曲线

Value model training results:

value_model曲线结果

**Step 4 — CFG Training**

CFG training curve:

截屏2026-04-10 02 12 07

CFG improvement histogram:

recap_libero10_task0

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/value_model/__init__.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            for k, v in state_dict.items()
+            if k.startswith("model.")
+        }
+        if len(set(stripped.keys()) & model_keys) > 0:
+            logger.info("Stripped 'model.' prefix from checkpoint keys")
+            return stripped
+    return state_dict
+
+
+def get_value_model(cfg: DictConfig, torch_dtype=None) -> ValueCriticModel:
```
Comment (qurakchin) [inline rlinf/models/embodiment/value_model/__init__.py]: use get_model?

File: examples/recap/process/config/model/value.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Value model config for inference
```
Comment (qurakchin) [inline examples/recap/process/config/model/value.yaml:1]: why use examples/process folder but not like "examples/recap_pi0.6" or "examples/recap"

File: examples/recap/process/config/model/value.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+critic_expert_variant: "gemma_1m"
+num_bins: 201
```
Comment (qurakchin) [inline examples/recap/process/config/model/value.yaml:3]: what's the num_bins? can you add a description of thi in the .rst

File: examples/process/config/compute_advantages.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#   cd examples/process && torchrun --nproc_per_node=4 compute_advantages.py \
+#       --config-name compute_advantages \
+#       advantage.value_checkpoint=/path/to/checkpoint
+
+defaults:
+  - model/value@advantage.model
+
+advantage:
+  value_checkpoint: /path/to/value_checkpoint
+  device: "cuda"
```
Comment (qurakchin) [inline examples/process/config/compute_advantages.yaml]: is the device needed? i see a line "cfg.advantage.device = device" and the line here may useless.

File: rlinf/workers/rollout/hf/huggingface_worker.py [hunk 5]
```diff
@@ -87,7 +88,19 @@ def __init__(self, cfg: DictConfig):
...
+        worker_seed = base_seed + int(self._rank)
+        seed_everything(worker_seed)
+        self.log_info(
+            "Initialized rollout RNG "
+            f"with seed={worker_seed} (base={base_seed}, rank={self._rank})"
+        )
+        return worker_seed
+
     def init_worker(self):
+        self._worker_seed = self._seed_worker_process()
```
Comment (qurakchin) [inline rlinf/workers/rollout/hf/huggingface_worker.py]: self._worker_seed is not used

File: rlinf/workers/rollout/hf/huggingface_worker.py [hunk 6]
```diff
@@ -87,7 +88,19 @@ def __init__(self, cfg: DictConfig):
         self.version = 0
         self.finished_episodes = None
 
+    def _seed_worker_process(self) -> int:
+        """Seed RNG state for this rollout worker process."""
+        base_seed = int(self.cfg.rollout.get("seed", self.cfg.actor.get("seed", 42)))
```
Comment (qurakchin) [inline rlinf/workers/rollout/hf/huggingface_worker.py]: should the seed be set even if the self.cfg.rollout has no attr seed.

File: rlinf/workers/cfg/fsdp_cfg_worker.py [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+                        data={
+                            "observation": observation,
+                            "actions": actions,
+                            "advantage": advantage,
+                        },
+                    )
+                    if isinstance(result, tuple) and len(result) == 2:
+                        losses, metrics_data = result
+                    else:
+                        losses = result
```
Comment (qurakchin) [inline rlinf/workers/cfg/fsdp_cfg_worker.py]: it's strange that loss = result?

File: examples/sft/config/model/value.yaml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+model_type: "value_model"
```
Comment (qurakchin) [inline examples/sft/config/model/value.yaml:1]: use a more meaningful name to differ from value_model.yaml

**Review by guozhen1997 — APPROVED** [review]

**Review by qurakchin — APPROVED** [review]

**Review by andylin-hao — APPROVED** [review]

**Comment by guozhen1997** [comment]
Need CI for CFG training.
I'm working on it.
